### PR TITLE
DBZ-768 Default values with variable second fractions

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-768
@gunnarmorling Please think about impact of timezones on default values. Do we have it covered?

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java [hunk 1]
```diff
@@ -108,9 +111,18 @@ private Object convertToLocalDate(Column column, String value) {
      * @return the converted value;
      */
     private Object convertToLocalDateTime(Column column, String value) {
-        if (ALL_ZERO_TIMESTAMP.equals(value) && column.isOptional()) return null;
-        if (ALL_ZERO_TIMESTAMP.equals(value)) value = EPOCH_TIMESTAMP;
-        String timestampFormat = timestampFormat(column.length());
+        if (ALL_ZERO_TIMESTAMP.matcher(value).matches() && column.isOptional()) return null;
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java]: Let's create the matcher just once by splitting up the condition. Same below.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/AbstractMysqlDefaultValueTest.java [hunk 2]
```diff
@@ -317,4 +320,36 @@ public void parseNumericAndDecimalToDecimalDefaultValue() {
...
+                "  G datetime NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+                "  H datetime NOT NULL DEFAULT '0000-00-00 00:00:00',\n" +
+                "  I datetime(3) NOT NULL DEFAULT '0000-00-00 00:00:00.000'," +
+                "  J datetime NOT NULL DEFAULT '2018-06-26 12:34:56'," +
+                "  K datetime(3) NOT NULL DEFAULT '2018-06-26 12:34:56.000'," +
+                "  L datetime(2) NOT NULL DEFAULT '2018-06-26 12:34:56.78'" +
+                ");";
+        parser.parse(sql, tables);
+        Table table = tables.forTable(new TableId(null, null, "TIME_TABLE"));
+        assertThat(table.columnWithName("A").defaultValue()).isEqualTo("1970-01-01T00:00:00+01:00");
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/AbstractMysqlDefaultValueTest.java]: @jpechane The expected value should be in UTC (i.e. no offset), as per the 0.8 docs:

> The `TIMESTAMP` type represents a timestamp without time zone information and is converted by MySQL from the server (or session's) current time zone into UTC when writing and vice versa when reading back the value.
Such columns are converted into an equivalent `io.debezium.time.ZonedTimestamp` in UTC based on the server (or session's) current time zone.
The timezone will be queried from the server by default.
If this fails, it must be specified explicitly as a connector option using the `database.serverTimezone` option.
So if for instance the database's time zone (either globally or configured for the connector by means of aforementioned option) is "America/Los_Angeles",
the `TIMESTAMP` value "2018-06-20 06:37:03" will be represented by a `ZonedTimestamp` with the value "2018-06-20T13:37:03Z".

Comment (gunnarmorling) [reply debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/AbstractMysqlDefaultValueTest.java]: The expecations for `DATETIME` below look alright.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/AbstractMysqlDefaultValueTest.java [hunk 3]
```diff
@@ -317,4 +320,36 @@ public void parseNumericAndDecimalToDecimalDefaultValue() {
         assertThat(table.columnWithName("B").defaultValue()).isEqualTo(BigDecimal.valueOf(2.321));
         assertThat(table.columnWithName("C").defaultValue()).isEqualTo(BigDecimal.valueOf(12.678));
     }
+
+    @Test
+    public void parseTimeDefaultValue() {
+        String sql = "CREATE TABLE TIME_TABLE (\n" +
+                "  A timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+                "  B timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',\n" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/AbstractMysqlDefaultValueTest.java]: Do we actually need the explicit line feeds? If so, let's have them for all lines consistently, although I don't think we need them at all.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java [hunk 4]
```diff
@@ -108,9 +109,19 @@ private Object convertToLocalDate(Column column, String value) {
...
     private Object convertToLocalDateTime(Column column, String value) {
-        if (ALL_ZERO_TIMESTAMP.equals(value) && column.isOptional()) return null;
-        if (ALL_ZERO_TIMESTAMP.equals(value)) value = EPOCH_TIMESTAMP;
-        String timestampFormat = timestampFormat(column.length());
+        final boolean matches = ALL_ZERO_TIMESTAMP.matcher(value).matches();
+        if (matches && column.isOptional()) return null;
+        if (matches) {
+            value = EPOCH_TIMESTAMP;
+            if (column.length() > 0) {
+                final StringBuilder sb = new StringBuilder(EPOCH_TIMESTAMP).append('.');
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java]: Why is it that the 0 are appended? A comment would be good. Also why are length 0's appended?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java [hunk 5]
```diff
@@ -124,8 +135,9 @@ private Object convertToLocalDateTime(Column column, String value) {
      * @return the converted value;
      */
     private Object convertToTimestamp(Column column, String value) {
-        if (ALL_ZERO_TIMESTAMP.equals(value) && column.isOptional()) return null;
-        if (ALL_ZERO_TIMESTAMP.equals(value)) value = EPOCH_TIMESTAMP;
+        final boolean matches = ALL_ZERO_TIMESTAMP.matcher(value).matches();
+        if (matches && column.isOptional()) return null;
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDefaultValuePreConverter.java]: Sorry for nit-picking, but let's not use the braceless if-style, I think it's very easy to be mis-read. I'd suggest the following here (and similarly above):

    if (matches) {
        if (column.isOptional()) {
             return null;
        }
        else {
            value = EPOCH_TIMESTAMP;
        }
    }

**Comment by jpechane** [comment]
@gunnarmorling Changes incorporated

**Comment by gunnarmorling** [comment]
Applied. Thanks, @jpechane!

**Pull request merged**